presto-main/src/main/java/com/facebook/presto/operator/ArrayBlockEncodingBuffers.java

```diff
@@ -0,0 +1,331 @@
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.operator.ByteArrayUtils.writeLengthPrefixedString;
+import static com.google.common.base.Verify.verify;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+import static sun.misc.Unsafe.ARRAY_INT_INDEX_SCALE;
+
+public class ArrayBlockEncodingBuffers
+        extends BlockEncodingBuffers
+{
+    private static final String NAME = "ARRAY";
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(ArrayBlockEncodingBuffers.class).instanceSize();
+
+    // The buffer for the offsets for all incoming blocks so far
+    private byte[] offsetsBuffer;
+
+    // The address that the next offset value will be written to.
+    private int offsetsBufferIndex;
+
+    // This array holds the condensed offsets for each position for the incoming block. It starts from the offset of row 1 instead of 0.
+    private int[] offsets;
+
+    // This array holds the offsets into its nested raw block for each top level row. It starts from the offset of row 1 instead of 0.
+    private int[] positionOffsets;
+
+    // The current incoming ArrayBlock is converted into ColumnarArray
+    private ColumnarArray columnarArray;
+
+    private final BlockEncodingBuffers rawBlockBuffer;
+
+    public ArrayBlockEncodingBuffers(DecodedBlockNode decodedBlockNode, int initialBufferPositions)
+    {
+        super(initialBufferPositions);
+        rawBlockBuffer = createBlockEncodingBuffers(decodedBlockNode.getChildren().get(0), initialBufferPositions);
+        prepareBuffers();
+    }
+
+    @Override
+    protected void prepareBuffers()
+    {
+        offsetsBuffer = new byte[initialBufferPositions * ARRAY_INT_INDEX_SCALE];
+        offsetsBufferIndex = 0;
+
+        rawBlockBuffer.prepareBuffers();
+    }
+
+    @Override
+    public void resetBuffers()
+    {
+        bufferedPositionCount = 0;
+        offsetsBufferIndex = 0;
+        resetNullsBuffer();
+
+        rawBlockBuffer.resetBuffers();
+    }
+
+    @Override
+    protected void setupDecodedBlockAndMapPositions(DecodedBlockNode decodedBlockNode)
+    {
+        requireNonNull(decodedBlockNode, "decodedBlockNode is null");
+
+        mapPositions(decodedBlockNode);
+        populateNestedPositions();
+        rawBlockBuffer.setupDecodedBlockAndMapPositions(decodedBlockNode.getChildren().get(0));
+    }
+
+    @Override
+    public void accumulateRowSizes(int[] rowSizes)
+    {
+        // Top level positionCount should be greater than 0
+        verify(positionCount > 0);
+
+        int averageElementSize = Integer.BYTES + Byte.BYTES;
+
+        for (int i = 0; i < positionCount; i++) {
+            rowSizes[i] += averageElementSize;
+        }
+
+        ensurePositionOffsetsSize();
+        System.arraycopy(offsets, 0, positionOffsets, 0, positionCount);
```

Comment:
This copy doesn't seem to belong here. I'd move this into `populateNestedPositions` where `offsets` are populated.